reconcile.go

```diff
@@ -232,56 +260,70 @@ func (a *Autopilot) pruneDeadServers() error {
 		return nil
 	}
 
-	state := a.GetState()
-
-	failed, voters, err := a.getFailedServers()
-	if err != nil || failed == nil {
+	servers, err := a.categorizeServers(a.promoter.PotentialVoterPredicate)
+	if err != nil {
 		return err
 	}
 
-	failed = a.promoter.FilterFailedServerRemovals(conf, state, failed)
+	state := a.GetState()
 
-	// remove failed non voting servers
-	for _, srv := range failed.FailedNonVoters {
-		a.logger.Info("Attempting removal of failed server node", "id", srv.ID, "name", srv.Name, "address", srv.Address)
-		a.delegate.RemoveFailedServer(srv)
+	// Support not breaking the promoter's interface for filtering servers
+	failedServers := servers.convertToFailedServers(state)
+	failedServers = a.promoter.FilterFailedServerRemovals(conf, state, failedServers)
+	servers = servers.convertFromFailedServers(failedServers)
+
+	// Partially apply (wrap) some functions for clarity below
+	removeStale := removeStale(a)
+	removeFailed := removeFailed(a)
+	adjudicate := func(logger hclog.Logger, minQuorum uint, voterCountProvider func() int) func(RaftServers) []raft.ServerID {
+		return func(raftServers RaftServers) []raft.ServerID {
+			return adjudicateRemoval(logger, minQuorum, voterCountProvider, raftServers)
+		}
+	}(a.logger, conf.MinQuorum, servers.PotentialVoters)
+
+	// Try to remove servers in order of increasing precedence
+
+	// Remove all stale non-voters
+	if err = remove(removeStale, adjudicate(servers.StaleNonVoters)); err != nil {
+		return err
 	}
 
-	// remove stale non voters
-	for _, id := range failed.StaleNonVoters {
-		a.logger.Debug("removing stale raft server from configuration", "id", id)
-		if err := a.removeServer(id); err != nil {
-			return err
-		}
+	// Remove stale voters
+	if err = remove(removeStale, adjudicate(servers.StaleVoters)); err != nil {
+		return err
 	}
```

Comment:
Similar to a comment above, I think it would be clearer if these things were implemented as methods on Autopilot instead of inlined/returned funcs that close on Autopilot.

I don't know why github put this comment on the old code as it really applies to the new code.